src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java

```diff
@@ -0,0 +1,432 @@
+import org.joda.time.DateTimeZone;
+import org.joda.time.MutableDateTime;
+import org.kuali.kfs.core.api.config.property.ConfigContext;
+import org.kuali.kfs.core.api.config.property.ConfigurationService;
+import org.kuali.kfs.core.api.datetime.DateTimeService;
+import org.kuali.kfs.krad.util.ObjectUtils;
+import org.kuali.kfs.krad.util.UrlFactory;
+import org.kuali.kfs.sys.KFSConstants;
+
+import edu.cornell.kfs.concur.ConcurConstants;
+import edu.cornell.kfs.concur.ConcurConstants.ConcurApiOperations;
+import edu.cornell.kfs.concur.ConcurConstants.ConcurApiParameters;
+import edu.cornell.kfs.concur.ConcurConstants.ConcurEventNoticationVersion2EventType;
+import edu.cornell.kfs.concur.ConcurConstants.ConcurEventNotificationVersion2ProcessingResults;
+import edu.cornell.kfs.concur.ConcurConstants.RequestV4Status;
+import edu.cornell.kfs.concur.ConcurConstants.RequestV4Views;
+import edu.cornell.kfs.concur.ConcurKeyConstants;
+import edu.cornell.kfs.concur.ConcurParameterConstants;
+import edu.cornell.kfs.concur.ConcurUtils;
+import edu.cornell.kfs.concur.batch.service.ConcurBatchUtilityService;
+import edu.cornell.kfs.concur.batch.service.ConcurEventNotificationV2WebserviceService;
+import edu.cornell.kfs.concur.batch.service.ConcurRequestV4Service;
+import edu.cornell.kfs.concur.businessobjects.ConcurAccountInfo;
+import edu.cornell.kfs.concur.businessobjects.ConcurEventNotificationProcessingResultsDTO;
+import edu.cornell.kfs.concur.businessobjects.ValidationResult;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4CustomItemDTO;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4ListItemDTO;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4ListingDTO;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4OperationDTO;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4ReportDTO;
+import edu.cornell.kfs.concur.rest.jsonObjects.ConcurRequestV4StatusDTO;
+import edu.cornell.kfs.concur.service.ConcurAccountValidationService;
+import edu.cornell.kfs.sys.CUKFSConstants;
+
+public class ConcurRequestV4ServiceImpl implements ConcurRequestV4Service {
+
+    private static final Logger LOG = LogManager.getLogger();
+
+    protected static final String PROCESSING_ERROR_MESSAGE = "Encountered an error while processing travel request";
+
+    protected ConcurBatchUtilityService concurBatchUtilityService;
+    protected ConcurEventNotificationV2WebserviceService concurEventNotificationV2WebserviceService;
+    protected ConcurAccountValidationService concurAccountValidationService;
+    protected ConfigurationService configurationService;
+    protected DateTimeService dateTimeService;
+
+    @Override
+    public List<ConcurEventNotificationProcessingResultsDTO> processTravelRequests(String accessToken) {
+        LOG.info("processTravelRequests, Starting processing of travel requests");
+        RequestV4QuerySettings querySettings = new RequestV4QuerySettings(
```

Comment:
I am finding this code hard to follow.  Perhaps it could use some refactoring to help with clarity.  My expectation for the logic flow is below.  Is my expectation too simplistic 

1. Get the travel requests in pending external validation based on the time timeframe logic you have figured out
2. If non-prod, filter the list for only test users, if prod, filter the list for only users that aren't test users
3. Loop through the list validating the accounting details
4. Accumulate ResultsDTO objects as you
5. Leave a holding place to update the status in concur on a future user story

Reply:
The steps you listed are the general logic that the current code is following. However, the filter-for-users step is a little more complex than merely filtering based on the approval user:

* When a request is Pending External Validation, the JSON doesn't show an active approver, and the prior approvers are not shown either.
* To determine which requests were approved by a test user, I add the test user's ID to the query parameters, and record the UUIDs of the resulting requests.
* To determine which requests should actually be processed, I run the query without specifying a user to filter on, and then I filter the results further based on whether or not they appear in the set of test-user-approved results.

I could simplify the is-test-request logic if there was some other easy-to-analyze field I could check. I could try checking the request's "owner" instead of the approver, as long as we can expect the test requests to only be created by test users. (I'm not familiar with Concur's user account setup, so I wasn't certain whether someone ever used Production accounts to create a test request.) If limiting based on owner is possible, then I could update the code to check for a known set of test users (Cheri, etc.).

Reply:
I like the idea of filtering by owner.  I expect that Tami is only going to test with Cher Knapp, or maybe a few other small subset of people.  I know Cheri will never have a production travel request.  We should check with Tami.